### DISCUSSION TITLE
querying sources that need auth

### DISCUSSION THREAD

**Discussion by mathib** [comunica/comunica#612] [Q&A] [answered]
#### Issue type:

- :question: Question

____
#### Question:

For some time, I'm wondering if Comunica could also be used to query both Linked Open Data and Linked Data Fragments that need authentication. My main use case would be to have a triplestore with authentication and some other public sources (RDF files, open SPARQL endpoints, TPF servers, etc.) that need to be queried (read) together. 

Is this something users of Comunica have to deal with in their own code somehow, or is this something planned in a future development? 

I can image that in the Solid project, there are similar use cases where one wants to query both closed and open Linked Data (not sure if it uses Comunica as a query engine)? As far as I understand, it's mainly focused on pods containing RDF files that require users to login to get read and/or write access. What is the method used here (if Comunica is used ofc)? I can imagine that after authentication, the user gets access to the RDF file and loads it into an in-memory store that can be queried together with other Linked Data Fragments (e.g. as in a previous question: https://github.com/comunica/comunica/issues/451). The problem is that this method is only suitable for RDF files that are not too large...

**Comment by rubensworks** [comment, marked as answer]
Adding authentication support to Comunica is pretty easy.
It's just a matter of replacing the default HTTP actor with one that can handle authenticated HTTP requests.

In the case of Solid, we have [`actor-http-solid-auth-fetch`](https://github.com/comunica/actor-http-solid-auth-fetch), which allows you to log in with Solid, and query over private (and public) files. This actor is not enabled in the default Comunica configurations, but it is for example using in [LDflex](https://github.com/LDflex/LDflex-Comunica).

Other forms of authentication can be implemented in a similar way as `actor-http-solid-auth-fetch`.

Does this answer your question?

**Comment by mathib** [comment]
Thanks for the pointers! Comunica has a large amount of modules, so it's not always that "easy" to find your way in it 😃 

Maybe it would be useful for other "dummies" like me to provide default options to connect easily (and safely) to a secured triplestore, e.g. using basic auth?

**Comment by rubensworks** [comment]
> Comunica has a large amount of modules, so it's not always that "easy" to find your way in it

True. A large-scale documentation effort will be starting soon... :-)

> Maybe it would be useful for other "dummies" like me to provide default options to connect easily (and safely) to a secured triplestore, e.g. using basic auth?

This would require the implementation of a new actor (like `actor-http-solid-auth-fetch`).
This is not something that is on our roadmap at the moment, but feel free to open a new feature request issue for this. Perhaps someone else is interested in picking this up.

**Comment by mathib** [comment]
> True. A large-scale documentation effort will be starting soon... :-)

Great!

> This would require the implementation of a new actor (like actor-http-solid-auth-fetch).

I'm currently lacking the time (and Comunica project knowledge) also... I'll make a new feature request. Thanks for the answers!
